feat: [PL-41822]: OpenApi spec for notification service api's

jennyjames21 opened the pull request:
## Describe your changes
Summary:
Centralized Notification Open Api Spec 

Test
-Local

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/54164)
<!-- Reviewable:end -->

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger uts

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+servers:
+  - url: 'https://app.harness.io'
+paths:
+  /notification-channel:
```
for Update Delete and Get calls - please use this path `/notification-channels/{notification-channel}`
for Create and List please use `/notification-channels`
and similar for notification-rule, thanks!

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+paths:
+  /notification-channel:
+    get:
+      summary: Your GET endpoint
```
Please add a small summary here

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/NotificationChannelDTO'
```
You would need to create a "response body" in stoplight and reference this model there, and the same thing for request bodies

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+        - schema:
+            type: string
+          in: query
+          name: accountIdentifier
```
accountIdentifier - this is no longer needed to be passed in Path param, please add the optional Header param for this
https://harness.atlassian.net/wiki/spaces/PLATFORM/pages/21203550744/Summary+of+API+Redesign+Changes#Account-Header-%26-Authentication

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+        - schema:
+            type: string
+          in: query
+          name: orgIdentifier
```
if your resource is scoped to project level then these need to be added to the path parameter as per REST standards. 
if this is present on multiple scopes then it will lead to different endpoints per scope 
eg. account secrets - v1/secrets
org secrets - v1/orgs/{org}/secrets
project secrets - v1/orgs/{org}/projects/{project}/secrets

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+      summary: ''
+      operationId: post-notification-channel
+      responses:
+        '200':
```
please remove the 200 response for post call

mankrit-singh replied:
it should be 201 instead of 200 when using create api, please update when possible

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+      summary: ''
+      operationId: delete-notification-channel
+      responses:
+        '200':
```
delete should have 204 response with no response body

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+      properties:
+        name:
+          type: string
+        orgIdentifier:
```
please use org, project (ignore Identifier) and use snake casing eg. notification_channel_type

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+        createdBy:
+          type: object
+          properties: {}
+        createdAt:
```
please use "created"

jennyjames21 replied:
updated

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,462 @@
+    NotificationRuleDTO:
+      type: object
+      properties:
+        notificationRuleIdentifier:
```
please fix casing for all parameters in models

jennyjames21 replied:
updated

mankrit-singh reviewed (commented):
Finally, please add a common Tag for each resource, the generated interfaces will be created based on this tag - 
eg.
https://github.com/harness/harness-core/blob/develop/platform-service/modules/resource-group-service/contracts/openapi/v1/openapi.yaml

For reference of API standards- https://harness.atlassian.net/wiki/spaces/PLATFORM/pages/21203550744/Summary+of+API+Redesign+Changes

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger ut5

jennyjames21 commented:
trigger ut7

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,581 @@
+          created_by: {}
+          created: 0
+      properties:
+        Identifier:
```
i in small

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,581 @@
+                        properties:
+                          name:
+                            type: string
+                          orgIdentifier:
```
could you make this "org" and "project" respectively?

jennyjames21 replied:
updated

mankrit-singh replied:
this is still orgIdentifier and projectIdentifier

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,581 @@
+servers:
+  - url: 'https://app.harness.io'
+paths:
+  'v1/orgs/{org}/projects/{project}/notification-channels':
```
also please start with with a slash 
'/v1/orgs/{org}/projects/{project}/notification-channels':

jennyjames21 replied:
updated

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,581 @@
+          - DESC
+      description: Order on the basis of which sorting is done.
+  responses:
+    NotificationChannelListResponse:
```
Will this also need Pagination? 
Point 2 here - https://harness.atlassian.net/wiki/spaces/PLATFORM/pages/21203550744/Summary+of+API+Redesign+Changes#API-Standards

jennyjames21 replied:
add to response class similar audit class

mankrit-singh commented on platform-service/modules/notification-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -0,0 +1,581 @@
+        name: project
+        in: path
+        required: true
+    get:
```
for every api, you would need to add x-api-key in "Security"

jennyjames21 replied:
updated

jennyjames21 commented:
trigger uts

mankrit-singh reviewed (commented):
You may also add the response bodies for Update and Create API, but otherwise the spec looks good (minor comments). Thanks!

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger uts

jennyjames21 commented:
> You may also add the response bodies for Update and Create API, but otherwise the spec looks good (minor comments). Thanks!

Added response for body for create and update call response to 201
Fix identifiers in example
Changed u

jennyjames21 commented:
trigger ut1

jennyjames21 commented:
trigger ut9

jennyjames21 commented:
trigger messagecheck

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger ut7

jennyjames21 commented:
trigger ut2

jennyjames21 commented:
trigger ut5

jennyjames21 commented:
trigger ut7

jennyjames21 commented:
trigger ut7

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger ut0

jennyjames21 commented:
trigger ut1

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger ut4

jennyjames21 commented:
trigger ut8

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger ut1

jennyjames21 commented:
trigger ut4

jennyjames21 commented:
trigger ut2

jennyjames21 commented:
trigger ut3

jennyjames21 commented:
trigger ut5

jennyjames21 commented:
trigger ut0

jennyjames21 commented:
trigger ut2

jennyjames21 commented:
trigger ut7

jennyjames21 commented:
trigger ut8

jennyjames21 commented:
trigger ut9

jennyjames21 commented:
trigget uts

jennyjames21 commented:
trigger uts

jennyjames21 commented:
trigger ut3
